x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/ShardChangesIT.java

```diff
@@ -681,6 +691,57 @@ public void testUnfollowIndex() throws Exception {
         assertThat(client().prepareSearch("index2").get().getHits().getTotalHits(), equalTo(2L));
     }
 
+    public void testFailOverOnFollower() throws Exception {
+        int numberOfReplicas = between(1, 2);
+        internalCluster().startMasterOnlyNode();
+        internalCluster().startDataOnlyNodes(numberOfReplicas + between(1, 2));
+        String leaderIndexSettings = getIndexSettings(1, numberOfReplicas,
+            singletonMap(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), "true"));
+        assertAcked(client().admin().indices().prepareCreate("leader-index").setSource(leaderIndexSettings, XContentType.JSON));
+        ensureGreen("leader-index");
+        AtomicBoolean stopped = new AtomicBoolean();
+        Thread[] threads = new Thread[between(1, 8)];
+        AtomicInteger docID = new AtomicInteger();
+        for (int i = 0; i < threads.length; i++) {
+            threads[i] = new Thread(() -> {
+                while (stopped.get() == false) {
+                    try {
+                        if (frequently()) {
+                            String id = Integer.toString(frequently() ? docID.incrementAndGet() : between(0, 10)); // sometimes update
+                            client().prepareIndex("leader-index", "doc", id).setSource("{\"f\":" + id + "}", XContentType.JSON).get();
+                        } else {
+                            String id = Integer.toString(between(0, docID.get()));
+                            client().prepareDelete("leader-index", "doc", id).get();
+                        }
+                    } catch (ElasticsearchException ignored) {
+
+                    }
+                }
+            });
+            threads[i].start();
+        }
+        PutFollowAction.Request follow = follow("leader-index", "follower-index");
+        client().execute(PutFollowAction.INSTANCE, follow).get();
+        ensureGreen("follower-index");
+        atLeastDocsIndexed("follower-index", between(20, 60));
+        final ClusterState clusterState = clusterService().state();
+        for (ShardRouting shardRouting : clusterState.routingTable().allShards("follower-index")) {
+            if (shardRouting.primary()) {
+                DiscoveryNode assignedNode = clusterState.nodes().get(shardRouting.currentNodeId());
+                internalCluster().restartNode(assignedNode.getName(), new InternalTestCluster.RestartCallback());
+                break;
+            }
+        }
+        ensureGreen("follower-index");
+        atLeastDocsIndexed("follower-index", between(80, 150));
+        stopped.set(true);
+        for (Thread thread : threads) {
+            thread.join();
+        }
+        assertSameDocCount("leader-index", "follower-index");
```

Comment:
can we strengthen this to actually check the docs and their seq#, terms and content?